crates/noirc_evaluator/src/ssa/conditional.rs

```diff
@@ -452,127 +495,202 @@ impl DecisionTree {
+                                let ins2 = ctx.get_mut_instruction(ins_id);
+                                ins2.operation = Operation::Binary(crate::node::Binary {
+                                    lhs: binop.lhs,
+                                    rhs: binop.rhs,
+                                    operator: binop.operator.clone(),
+                                    predicate: Some(ass_value),
+                                });
+                            }
                         }
                         _ => (),
                     }
                 }
-            }
-            Operation::Store { array_id, index, value } => {
-                if !ins.operation.is_dummy_store()
-                    && ctx.under_assumption(ass_value)
-                    && stack.created_arrays[array_id] != stack.block
-                {
-                    let load = Operation::Load { array_id: *array_id, index: *index };
-                    let e_type = ctx.mem[*array_id].element_type;
-                    let dummy =
-                        ctx.add_instruction(Instruction::new(load, e_type, Some(stack.block)));
-                    let operation = Operation::Cond {
-                        condition: ass_value,
-                        val_true: *value,
-                        val_false: dummy,
-                    };
-                    let cond =
-                        ctx.add_instruction(Instruction::new(operation, e_type, Some(stack.block)));
-
-                    stack.push(dummy);
-                    stack.push(cond);
-                    //store the conditional value
-                    let ins2 = ctx.get_mut_instruction(ins_id);
-                    ins2.operation =
-                        Operation::Store { array_id: *array_id, index: *index, value: cond };
-                }
-                stack.push(ins_id);
-            }
-            Operation::Intrinsic(_, _) => {
-                stack.push(ins_id);
-                if ctx.under_assumption(ass_value) {
-                    if let ObjectType::Pointer(a) = ins.res_type {
-                        if stack.created_arrays[&a] != stack.block {
-                            let array = &ctx.mem[a].clone();
-                            let name = array.name.to_string() + DUPLICATED;
-                            ctx.new_array(&name, array.element_type, array.len, None);
-                            let array_dup = ctx.mem.last_id();
+                Operation::Store { array_id, index, value } => {
+                    if !ins.operation.is_dummy_store() {
+                        if let Some(idx) = ctx.get_as_constant(*index) {
+                            if (idx.to_u128() as u32) >= ctx.mem[*array_id].len {
+                                if !DecisionTree::short_circuit(ctx, stack, ass_value) {
+                                    unreachable!(
+                                        "index out of bounds: the len is {} but the index is {}",
```

Comment:
This error message doesn't seem to match the failure to short-circuit

Reply:
Short-circuit makes sense only if the condition can be false, if not, we know that the circuit will fail and we are in the case where the index is out of bounds. So the error message is correct.

Reply:
I think my confusion is from thinking this was an internal error with ids due to the unreachable call, rather than an external error users can run into when using indices that are too large in noir code. We should never report user errors via `unreachable` or panics. I'll open an issue for it since this already happens in master.